typing/typecore.ml

```diff
@@ -5817,23 +5888,41 @@ and type_cases
      is to typecheck the guards and the cases, and then to check for some
      warnings that can fire in the presence of guards.
   *)
-  map_half_typed_cases category env ty_arg ty_res loc caselist ~check_if_total
+  map_half_typed_cases ?conts category env ty_arg ty_res loc caselist
+    ~check_if_total
     ~type_body:begin
-      fun { pc_guard; pc_rhs } pat ~ext_env ~ty_expected ~ty_infer
+      fun { pc_guard; pc_rhs } pat ~ext_env ~cont ~ty_expected ~ty_infer
           ~contains_gadt:_ ->
+        let cont, ext_env' =
+          match cont with
+          | Some (id, desc) ->
+              let ext_env =
+                Env.add_value ~check:(fun s -> Warnings.Unused_var_strict s)
+                  id desc ext_env
+              in
+                Some id, ext_env
+          | None -> None, ext_env
+        in
```

Comment:
Adding the continuation pattern variable at this point breaks the detection of multiply bound pattern variables.
For instance,
```ocaml
effect E k, k -> ignore k
```
is accepted.  The sub-Pr https://github.com/avsm/ocaml/pull/4 proposes to fix issue by adding the continuation variable to the pattern typing environment. This allows to both detect multiple binding for the continuation variable and move the split between the `when` and `case body` environment to the `map_half_typed` function where it feels slightly more natural to me.